### PR TITLE
Add support for absolute value failure

### PR DESCRIPTION
**Pull request by rtfpessoa**
This should be a skeleton of what we need. Still missing the tests.

### CODE REVIEW

File: src/main/scala/com/codacy/analysis/cli/Main.scala [hunk 1]
```diff
@@ -38,40 +39,65 @@ class MainImpl extends CLIApp {
...
-    }
 
-    Await.result(res, Duration.Inf) match {
-      case Left(e) =>
-        logger.error(s"Failed analysis: $e")
-      case Right(_) =>
-        logger.info("Completed analysis successfully")
+        val uploadResult = uploadResults(codacyClientOpt)(analyse.uploadValue, analyse.commitUuid, analysisResults)
+        Await.result(uploadResult, Duration.Inf) match {
+          case Left(err) if analyse.uploadValue =>
```
Comment (xplosunn) [inline src/main/scala/com/codacy/analysis/cli/Main.scala]: We can move this "if" to before the match, so we don't even run the "uploadResults" function if "analyse.uploadValue" is false.

File: src/main/scala/com/codacy/analysis/cli/Main.scala [hunk 2]
```diff
@@ -38,40 +39,65 @@ class MainImpl extends CLIApp {
...
     }
 
     ()
   }
 
+  private def uploadResults(codacyClientOpt: Option[CodacyClient])(
+    upload: Boolean,
+    commitUuid: Option[String],
+    executorResultsEither: Either[String, Seq[ExecutorResult]]): Future[Either[String, Unit]] = {
+    val resultsUploader: Either[String, ResultsUploader] = codacyClientOpt.fold {
```
Comment (xplosunn) [inline src/main/scala/com/codacy/analysis/cli/Main.scala]: I would prefer this as a for comprehension with the "Option"s converted to "Either"s using ".toRight", but not a mandatory change.

File: src/main/scala/com/codacy/analysis/cli/analysis/Status.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+  object ExitCodes {
+    val success = 0
+    val failedAnalysis = 1
+    val partiallyFailedAnalysis = 2
+    val maxAllowedIssuesExceeded = 101
+  }
+}
+
+class Status(maxAllowedIssues: Int, failIfIncomplete: Boolean = false) {
```
Comment (xplosunn) [inline src/main/scala/com/codacy/analysis/cli/analysis/Status.scala]: Highly dislike this class name. Maybe SuccessCriteria? Or something else that points to this being an optional failure threshold / configuration / settings.

File: src/main/scala/com/codacy/analysis/cli/command/CLIApp.scala [hunk 4]
```diff
@@ -114,8 +114,13 @@ final case class Analyse(
   upload: Int @@ Counter = Tag.of(0),
   @ExtraName("p") @ValueDescription("The number of tools to run in parallel")
   parallel: Option[Int] = Option.empty,
+  @ValueDescription("The maximum number of issues allowed to pass the analysis")
+  maxAllowedIssues: Int = 0,
```
Comment (xplosunn) [inline src/main/scala/com/codacy/analysis/cli/command/CLIApp.scala]: It makes sense for me that this value is optional, instead of failing by default on any issue found.

File: src/main/scala/com/codacy/analysis/cli/command/analyse/AnalyseExecutor.scala [hunk 5]
```diff
@@ -143,7 +93,7 @@ class AnalyseExecutor(toolInput: Option[String],
         logger.error(s"""Could not find conditions to run tool ${tool.name} with:
              |shouldUseConfigFile:$shouldUseConfigFile = notEdited:${toolConfiguration.notEdited} && foundToolConfigFile:$hasConfigFiles
              |shouldUseRemoteConfiguredPatterns:$shouldUseRemoteConfiguredPatterns = !shouldUseConfigFile:$shouldUseConfigFile && allowsUIConfiguration:${tool.allowsUIConfiguration} && hasPatterns:${toolConfiguration.patterns.nonEmpty}
-             |shouldRun:$shouldRun = !needsPatternsToRun:${tool.needsPatternsToRun} || shouldUseConfigFile:$shouldUseConfigFile || shouldUseRemoteConfiguredPatterns:$shouldUseRemoteConfiguredPatterns
+             |shouldRun:$shouldRun = !needsPatternsToRun:${tool.needsPatternsToRun}|| shouldUseConfigFile:$shouldUseConfigFile|| shouldUseRemoteConfiguredPatterns:$shouldUseRemoteConfiguredPatterns
```
Comment (xplosunn) [inline src/main/scala/com/codacy/analysis/cli/command/analyse/AnalyseExecutor.scala]: Why this change?

File: src/main/scala/com/codacy/analysis/cli/utils/EitherOps.scala [hunk 6]
```diff
@@ -0,0 +1,34 @@
+package com.codacy.analysis.cli.utils
+
+import scala.util.{Failure, Success, Try}
+
+object EitherOps {
+
+  def sequenceWithFixedLeftUnit[A](left: A)(eitherIterable: Seq[Either[A, Unit]]): Either[A, Unit] = {
```
Comment (xplosunn) [inline src/main/scala/com/codacy/analysis/cli/utils/EitherOps.scala]: "sequenceUnitWithFixedLeft" makes more sense to me, if you want to add "Unit" to the name.

File: src/main/scala/com/codacy/analysis/cli/utils/EitherOps.scala [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    eitherIterable
+      .foldLeft[Either[A, Unit]](Right(())) { (acc, either) =>
+        acc.flatMap(_ => either)
+      }
+      .left
+      .map(_ => left)
+  }
+
+  def sequenceWithFixedLeft[A, B](left: A)(eitherIterable: Set[Either[A, B]]): Either[A, Set[B]] = {
```
Comment (xplosunn) [inline src/main/scala/com/codacy/analysis/cli/utils/EitherOps.scala]: My mistake probably, but rename "eitherIterable" => "eitherSet"

File: src/main/scala/com/codacy/analysis/cli/utils/EitherOps.scala (hunk 6 above)
Comment (xplosunn) [inline src/main/scala/com/codacy/analysis/cli/utils/EitherOps.scala]: My mistake probably, but rename "eitherIterable" => "eitherSeq"

File: src/main/scala/com/codacy/analysis/cli/utils/EitherOps.scala [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+      .foldLeft[Either[A, Set[B]]](Right(Set.empty)) { (acc, either) =>
+        acc.flatMap { bSeq =>
+          either.map(b => bSeq ++ Set(b))
+        }
+      }
+      .left
+      .map(_ => left)
+  }
+
+  def fromTry[A](tryValue: Try[A]): Either[String, A] = {
```
Comment (xplosunn) [inline src/main/scala/com/codacy/analysis/cli/utils/EitherOps.scala]: This seems redundant with TryOps.toRight
(if you don't want to change the existing one we can have another one that has Throwable => L as the argument)

File: src/main/scala/com/codacy/analysis/cli/command/CLIApp.scala (hunk 4 above)
Comment (rtfpessoa) [reply src/main/scala/com/codacy/analysis/cli/command/CLIApp.scala]: This is the same default as the cloud quality settings

File: src/main/scala/com/codacy/analysis/cli/command/analyse/AnalyseExecutor.scala (hunk 5 above)
Comment (rtfpessoa) [reply src/main/scala/com/codacy/analysis/cli/command/analyse/AnalyseExecutor.scala]: Automatic formatter ?!?

File: src/main/scala/com/codacy/analysis/cli/analysis/Status.scala [hunk 9]
```diff
@@ -0,0 +1,39 @@
+package com.codacy.analysis.cli.analysis
+
+import com.codacy.analysis.cli.command.analyse.AnalyseExecutor.ExecutorResult
+
+object Status {
+
+  object ExitCodes {
```
Comment (xplosunn) [inline src/main/scala/com/codacy/analysis/cli/analysis/Status.scala]: Add these to the README?

File: src/main/scala/com/codacy/analysis/cli/analysis/Status.scala [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    val success = 0
+    val failedAnalysis = 1
+    val partiallyFailedAnalysis = 2
+    val failedUpload = 101
+    val maxAllowedIssuesExceeded = 201
+  }
+}
+
+class Status(maxAllowedIssues: Int, failIfIncomplete: Boolean = false) {
```
Comment (xplosunn) [inline src/main/scala/com/codacy/analysis/cli/analysis/Status.scala]: This class could use a better name, because it's not the status of the CLI.

File: src/main/scala/com/codacy/analysis/cli/command/analyse/AnalyseExecutor.scala [hunk 11]
```diff
@@ -128,10 +81,14 @@ class AnalyseExecutor(toolInput: Option[String],
                                    remoteConfiguration: Either[String, ProjectConfiguration]): Try[Configuration] = {
     val (baseSubDir, extraValues) = getExtraConfiguration(localConfiguration, tool)
     (for {
+      // scalafix:off NoInfer.product
```
Comment (xplosunn) [inline src/main/scala/com/codacy/analysis/cli/command/analyse/AnalyseExecutor.scala]: Does it work if you add the type annotation instead of this scalafix comment?

File: src/main/scala/com/codacy/analysis/cli/upload/ResultsUploader.scala [hunk 12]
```diff
@@ -1,13 +1,40 @@
...
+
+  private val logger: Logger = getLogger
+
+  def apply(codacyClientOpt: Option[CodacyClient],
+            upload: Boolean,
+            commitUuid: Option[String],
+            batchSize: Option[Int] = Option.empty[Int])(
+    implicit context: ExecutionContext): Either[String, Option[ResultsUploader]] = {
+    if (upload) {
+      codacyClientOpt.fold {
```
Comment (xplosunn) [inline src/main/scala/com/codacy/analysis/cli/upload/ResultsUploader.scala]: I'd favour a for comprehension here, instead of a nested fold.

File: src/main/scala/com/codacy/analysis/cli/command/analyse/AnalyseExecutor.scala (hunk 11 above)
Comment (rtfpessoa) [reply src/main/scala/com/codacy/analysis/cli/command/analyse/AnalyseExecutor.scala]: Give it a try, I could not make it work.

This was already disabled, but I decided to revert it and disable only the rule in this spot.

File: src/main/scala/com/codacy/analysis/cli/upload/ResultsUploader.scala (hunk 12 above)
Comment (rtfpessoa) [reply src/main/scala/com/codacy/analysis/cli/upload/ResultsUploader.scala]: The for would make the condition less obvious.

This way when reading, you know exactly that nothing will go on for the condition. In the for you would have stuff before that would never execute and would just cause confusion.
